File: frontend/src/service/table/Order.ts
import { z } from 'zod'

export const OrderVariantSchema = z.object({
  id: z.number().int().min(1),
  name: z.string().min(1).max(100),
  priceCents: z.number().int().min(0),
  quantity: z.number().int().min(1),
})
export type OrderVariant = z.infer<typeof OrderVariantSchema>

export const PlaceOrderSchema = z.object({
  tableId: z.number().int().min(1),
  variants: OrderVariantSchema.array().min(1),
  comment: z.string().max(100),
})
export type PlaceOrder = z.infer<typeof PlaceOrderSchema>

export const OrderSchema = z.object({
  id: z.uuid(),
  userId: z.number().int().min(1),
  tableId: z.number().int().min(1),
  variants: OrderVariantSchema.array().min(1),
  totalPriceCents: z.number().int().min(0),
  comment: z.string().max(100),
  placedAt: z.string().refine((date) => !isNaN(Date.parse(date)), {
    message: 'Invalid date format',
  }),
})
export type Order = z.infer<typeof OrderSchema>

File: backend/domain/table/order.go
package table

import (
	"time"

	z "github.com/Oudwins/zog"
	"github.com/nicograef/jotti/backend/domain/product"
)

type OrderVariant struct {
	ID         int    `json:"id"`
	Name       string `json:"name"`
	PriceCents int    `json:"priceCents"`
	Quantity   int    `json:"quantity"`
}

var orderVariantSchema = z.Struct(z.Shape{
	"ID":         product.IDSchema.Required(),
	"Name":       product.NameSchema.Required(),
	"PriceCents": product.PriceCentsSchema.Required(),
	"Quantity":   z.Int().GTE(1, z.Message("Quantity must be at least 1")).Required(),
})

type Order struct {
	ID              string         `json:"id"`
	UserID          int            `json:"userId"`
	TableID         int            `json:"tableId"`
	Variants        []OrderVariant `json:"variants"`
	TotalPriceCents int            `json:"totalPriceCents"`
	Comment         string         `json:"comment"`
	PlacedAt        time.Time      `json:"placedAt"`
}

var orderSchema = z.Struct(z.Shape{
	"ID":              z.String().UUID().Required(),
	"UserID":          z.Int().GTE(1).Required(),
	"TableID":         z.Int().GTE(1).Required(),
	"Variants":        z.Slice(orderVariantSchema).Min(1).Required(),
	"TotalPriceCents": z.Int().GTE(0).Required(),
	"Comment":         z.String().Max(100),
	"PlacedAt":        z.Time().Required(),
})
